modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationSerializationUtil.java

```diff
@@ -0,0 +1,414 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.configuration.util;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Utility class to help serializing  deserializing configuration values - primitives, strings, or arrays of primitives or strings.
+ */
+public class ConfigurationSerializationUtil {
+    private static final byte BOOLEAN = 1;
+
+    private static final byte BYTE = 2;
+
+    private static final byte SHORT = 3;
+
+    private static final byte INT = 4;
+
+    private static final byte LONG = 5;
+
+    private static final byte CHAR = 6;
+
+    private static final byte FLOAT = 7;
+
+    private static final byte DOUBLE = 8;
+
+    private static final byte STRING = 9;
+
+    private static final byte ARRAY = (byte) 0x80;
+
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
+
+    /**
+     * Converts a configuration value into a byte array.
+     *
+     * @param value Configuration values.
+     * @return Serialized value.
+     */
+    public static byte[] toBytes(Object value) {
+        Objects.requireNonNull(value);
+
+        byte header = header(value.getClass());
+
+        switch (header) {
+            case BOOLEAN:
+                return new byte[]{header, (byte) ((boolean) value ? 1 : 0)};
+
+            case BYTE:
+                return new byte[]{header, (byte) value};
+
+            case SHORT:
+                return allocateBuffer(Short.BYTES + 1).put(header).putShort((short) value).array();
+
+            case INT:
+                return allocateBuffer(Integer.BYTES + 1).put(header).putInt((int) value).array();
+
+            case LONG:
+                return allocateBuffer(Long.BYTES + 1).put(header).putLong((long) value).array();
+
+            case CHAR:
+                return allocateBuffer(Character.BYTES + 1).put(header).putChar((char) value).array();
+
+            case FLOAT:
+                return allocateBuffer(Float.BYTES + 1).put(header).putFloat((float) value).array();
+
+            case DOUBLE:
+                return allocateBuffer(Double.BYTES + 1).put(header).putDouble((double) value).array();
+
+            case STRING: {
+                byte[] strBytes = ((String) value).getBytes(StandardCharsets.UTF_8);
+
+                return allocateBuffer(1 + strBytes.length).put(header).put(strBytes).array();
+            }
+
+            case BOOLEAN | ARRAY: {
+                boolean[] booleans = (boolean[]) value;
+
+                int payloadSize = (booleans.length + Byte.SIZE - 1) / Byte.SIZE;
+
+                ByteBuffer buf = allocateBuffer(2 + payloadSize);
+
+                buf.put(header);
+
+                buf.put((byte) (payloadSize * Byte.SIZE - booleans.length));
```

Comment:
I suggest to extract `payloadSize * Byte.SIZE - booleans.length` and name the variable like `paddingLength` to make it clear for the reader what it is.